libraries/SITL/SIM_StateSpace.cpp

```diff
@@ -0,0 +1,547 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  State Space simulator class
+*/
+
+#include "SIM_StateSpace.h"
+#include <AP_Filesystem/AP_Filesystem.h>
+
+#include <stdio.h>
+#include <sys/stat.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Logger/AP_Logger.h>
+
+namespace SITL {
+
+StateSpace::StateSpace(const char *frame_str) :
+    Aircraft(frame_str)
+{
+
+    
+    if (strstr(frame_str, "-heli")) {
+        frame_type = HELI_FRAME;
+
+    } else if (strstr(frame_str, "-multi")) {
+        frame_type = MULTI_FRAME;
+
+    } else {
+        frame_type = MULTI_FRAME;
+    }
+
+    const char *colon = strchr(frame_str, ':');
+    size_t slen = strlen(frame_str);
+    if (colon != nullptr && slen > 5 && strcmp(&frame_str[slen-5], ".json") == 0) {
+        load_frame_params(colon+1);
+    }
+
+    ground_behavior = GROUND_BEHAVIOR_NO_MOVEMENT;
+    lock_step_scheduled = true;
+
+    motor_mask |= (1U<<0);
+}
+
+/*
+  update the helicopter simulation by one time step
+ */
+void StateSpace::update(const struct sitl_input &input)
+{
+
+    static uint64_t last_calc_us;
+    uint64_t now_us = AP_HAL::micros64();
+    float dt = 0.0f;
+    if (last_calc_us != 0) {
+        dt = (now_us - last_calc_us)*1.0e-6;
+    }
+    last_calc_us = now_us;
+
+    // get wind vector setup
+    update_wind(input);
+
+    Vector3f rot_accel;
+
+    switch (frame_type) {
+        case HELI_FRAME: {
+            if (model.coll_max - model.coll_min > 0.0f) {
+                hover_coll = (model.coll_hover - model.coll_min) / (model.coll_max - model.coll_min);
+            }
+            float servos_raw[16];
```

Comment:
Is there a define we could use for the max servos? I thought certain versions of AP support 32.